bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -5190,14 +5204,17 @@ void ReplicaImp::finalizeExecution() {
     auto epoch = bftEngine::EpochManager::instance().getSelfEpochNumber();
     bftEngine::EpochManager::instance().setSelfEpochNumber(epoch);
     bftEngine::EpochManager::instance().setGlobalEpochNumber(epoch);
-    Digest checkDigest;
-    Digest otherDigest;
+    Digest stateDigest, reservedPagesDigest, rvbDataDigest;
     CheckpointMsg::State checkState;
     const uint64_t checkpointNum = lastExecutedSeqNum / checkpointWindowSize;
-    stateTransfer->getDigestOfCheckpoint(
-        checkpointNum, sizeof(Digest), checkState, (char *)&checkDigest, (char *)&otherDigest);
-    CheckpointMsg *checkMsg =
-        new CheckpointMsg(config_.getreplicaId(), lastExecutedSeqNum, checkState, checkDigest, otherDigest, false);
+    stateTransfer->getDigestOfCheckpoint(checkpointNum,
+                                         sizeof(Digest),
+                                         checkState,
+                                         stateDigest.content(),
+                                         reservedPagesDigest.content(),
+                                         rvbDataDigest.content());
+    CheckpointMsg *checkMsg = new CheckpointMsg(
```

Comment:
when is checkMsg is deleted. I think as new and delete callers are different, better we use std::shared_ptr to avoid any memory leak.

Reply:
I'm afraid that this is out of the scope for his PR, as this requires a large change.

This checkpoint is transffered to multiple modules as a raw pointer. The big change here is to hold the internal pointer as shared one (since the object is pointed by multiple modules), and not the one outside.